File: lib/features/product/domain/repositories/product_repository.dart
import 'package:dartz/dartz.dart';

import '../../../../core/errors/failures.dart';
import '../../../../core/pagination/pagination.dart';
import '../entities/product_entity.dart';

/// Domain repository contract for Product. Implementations live in the
/// `data/` layer.
abstract class ProductRepository {
  Future<Either<Failure, List<ProductEntity>>> getAll();
  Future<Either<Failure, PaginatedResponse<ProductEntity>>> getAllPaginated(
    PaginationParams params,
  );
  Future<Either<Failure, ProductEntity>> getById(String id);
  Future<Either<Failure, ProductEntity>> add(ProductEntity entity);
  Future<Either<Failure, ProductEntity>> update(ProductEntity entity);
  Future<Either<Failure, Unit>> delete(String id);
  Future<Either<Failure, List<ProductEntity>>> search(String query);
}

File: packages/example_react_app/src/features/product/domain/repositories/product.repository.ts
/// Repository contract for Product.
import type { Either } from '../../../../core/usecase/usecase';
import type { Failure } from '../../../../core/errors/failures';
import type { ProductEntity } from '../entities/product.entity';
import type { PaginationParams, PaginatedResponse } from '../../../../core/pagination/pagination';

export interface ProductRepository {
  getAll(): Promise<Either<Failure, ProductEntity[]>>;
  getById(id: string): Promise<Either<Failure, ProductEntity>>;
  add(entity: ProductEntity): Promise<Either<Failure, ProductEntity>>;
  update(entity: ProductEntity): Promise<Either<Failure, ProductEntity>>;
  delete(id: string): Promise<Either<Failure, void>>;
  search(query: string): Promise<Either<Failure, ProductEntity[]>>;
  getAllPaginated(params: PaginationParams): Promise<Either<Failure, PaginatedResponse<ProductEntity>>>;
}
